### PR TITLE
Handles email mismatch case and fixes bug for federated credential already in use.

### PR DESCRIPTION
**Pull request by protocol86**
Handles email mismatch case and fixes bug for federated credential already in use.

### CODE REVIEW

File: FirebaseAuthUI/FUIAuth.m [hunk 1]
```diff
@@ -31,6 +31,18 @@
...
 
+/** @typedef EmailHintSignInCallback
+    @brief The type of block invoked when an emailHint sign-in event completes.
+
+    @param authResult Optionally; Result of sign-in request containing both the user and
+       the additional user info associated with the user.
+    @param error Optionally; the error which occurred - or nil if the request was successful.
+    @param credential Optionally; The credential used to sign-in.
+ */
+typedef void (^EmailHintSignInCallback)(FIRAuthDataResult *_Nullable authResult,
```
Comment (morganchen12) [inline FirebaseAuthUI/FUIAuth.m]: Should this be `FUIEmailHintSignInCallback`?

File: FirebaseAuthUI/FUIAuth.m [hunk 2]
```diff
@@ -299,7 +325,7 @@ - (void)signInWithEmailHint:(NSString *)emailHint
   [self.auth fetchProvidersForEmail:emailHint completion:^(NSArray<NSString *> *_Nullable providers,
                                                            NSError *_Nullable error) {
     if (error) {
-      completion(nil, error);
+      completion(nil, error, nil);
```
Comment (morganchen12) [inline FirebaseAuthUI/FUIAuth.m]: Is `completion` guaranteed to be nonnull here?

File: FirebaseAuthUI/FUIAuth.m [hunk 3]
```diff
@@ -351,16 +381,49 @@ - (void)signInWithEmailHint:(NSString *)emailHint
...
+              }
+
+              // Handle potential email mismatch.
+              if (![emailHint isEqualToString:authResult.user.email]) {
+                NSString *signedInEmail = authResult.user.email;
+                NSString *title =
+                    [NSString stringWithFormat:@"Continue sign in with %@?", signedInEmail];
+                NSString *message =
+                    [NSString stringWithFormat:@"You originally wanted to sign in with %@",
+                    emailHint];
```
Comment (morganchen12) [inline FirebaseAuthUI/FUIAuth.m]: If these strings are user-facing, they should be localized.

File: FirebaseAuthUI/FUIAuth.m [hunk 4]
```diff
@@ -351,16 +381,49 @@ - (void)signInWithEmailHint:(NSString *)emailHint
...
+                [FUIAuthBaseViewController showAlertWithTitle:title
+                                                      message:message
+                                                  actionTitle:@"Continue"
```
Comment (morganchen12) [inline FirebaseAuthUI/FUIAuth.m]: This string also requires localization.

File: FirebaseAuthUI/FUIAuth.m (hunk 1 above)
Comment (protocol86) [reply FirebaseAuthUI/FUIAuth.m]: I didn't think we need the prefix here since it's an internal callback.

File: FirebaseAuthUI/FUIAuth.m (hunk 3 above)
Comment (protocol86) [reply FirebaseAuthUI/FUIAuth.m]: The entire flow needs to be localized in a follow up CL.

File: FirebaseAuthUI/FUIAuth.m (hunk 4 above)
Comment (protocol86) [reply FirebaseAuthUI/FUIAuth.m]: Will localize in follow up CL.

File: FirebaseAuthUI/FUIAuth.m (hunk 2 above)
Comment (protocol86) [reply FirebaseAuthUI/FUIAuth.m]: Good catch, fixed.

**Pull request merged**